### PR TITLE
[SCB-2330] add injectFault into governance

### PR DESCRIPTION
**Pull request by zhaozw**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    return "servicecomb.faultInjection." + policy.getName();
+  }
+
+  @Override
+  public FaultInjectionPolicy matchPolicy(GovernanceRequest governanceRequest) {
+    return matchersManager.match(governanceRequest, faultInjectionProperties.getParsedEntity());
+  }
+
+  @Override
+  protected FaultInjectionPolicy createProcessor(String key, GovernanceRequest governanceRequest,
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java:46]: Should here return Fault based on policy?

Comment (liubao68) [reply governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java:46]: I think it's better create a test case like FlowControlTest to demonstrator how to use this.

Comment (zhaozw) [reply governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java:46]: OK

File: governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  private FaultInjectionProperties faultInjectionProperties;
+
+  public FaultInjectionHandler(FaultInjectionProperties faultInjectionProperties) {
+    this.faultInjectionProperties = faultInjectionProperties;
+  }
+
+  @Override
+  protected String createKey(GovernanceRequest governanceRequest, FaultInjectionPolicy policy) {
+    return "servicecomb.faultInjection." + policy.getName();
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java:36]: use constant `MATCH_FAULT_INJECTION_KEY` in properties class

File: governance/src/test/java/org/apache/servicecomb/governance/FlowControlTest.java [hunk 3]
```diff
@@ -21,8 +21,10 @@
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 
+import org.apache.servicecomb.governance.handler.FaultInjectionHandler;
```
Comment (liubao68) [inline governance/src/test/java/org/apache/servicecomb/governance/FlowControlTest.java:24]: create tests in a new class, e.g.  FaultInjectionHandlerTest

File: governance/src/test/java/org/apache/servicecomb/governance/FlowControlTest.java [hunk 4]
```diff
@@ -123,4 +132,93 @@ public void test_rate_limiting_service_name_work() throws Throwable {
...
+    Assertions.assertEquals("test", ds.get());
+
+    // flow control
+    CountDownLatch cd = new CountDownLatch(10);
+    AtomicBoolean notExpected = new AtomicBoolean(false);
+    AtomicBoolean delayExpected = new AtomicBoolean(false);
+    AtomicBoolean abortExpected = new AtomicBoolean(false);
+    for (int i = 0; i < 10; i++) {
+      new Thread(() -> {
+        fault.injectFault(faultResponse -> {
```
Comment (liubao68) [inline governance/src/test/java/org/apache/servicecomb/governance/FlowControlTest.java:200]: From the usage, seems the users need to call Fault api himself. This is not a correct usage case. 

see FlowControlTests for usage exampes. We should call ds.get and expecting errors.

File: governance/src/test/java/org/apache/servicecomb/governance/FaultInjectionTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  public void setFaultInjectionHandler(FaultInjectionHandler faultInjectionHandler) {
+    this.faultInjectionHandler = faultInjectionHandler;
+  }
+
+  public FaultInjectionTest() {
+  }
+
+  @Test
+  public void test_delay_fault_injection_service_name_work() throws Throwable {
+    DecorateCheckedSupplier<Object> dcs = Decorators.ofCheckedSupplier(() -> "test");
```
Comment (liubao68) [inline governance/src/test/java/org/apache/servicecomb/governance/FaultInjectionTest.java:52]: Maybe  FaultInjectionDecorators create a FaultInjectionDecorateCheckedSupplier is simpler

File: governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.apache.servicecomb.governance.marker.GovernanceRequest;
+import org.apache.servicecomb.governance.policy.FaultInjectionPolicy;
+import org.apache.servicecomb.governance.properties.FaultInjectionProperties;
+import org.apache.servicecomb.injection.Fault;
+import org.apache.servicecomb.injection.FaultInjectionUtil;
+
+public class FaultInjectionHandler extends AbstractGovernanceHandler<Fault, FaultInjectionPolicy> {
+
+  private FaultInjectionProperties faultInjectionProperties;
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/handler/FaultInjectionHandler.java:28]: This field can be final

File: governance/src/main/java/org/apache/servicecomb/governance/policy/FaultInjectionPolicy.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+package org.apache.servicecomb.governance.policy;
+
+import java.time.Duration;
+
+public class FaultInjectionPolicy extends AbstractPolicy {
+  public static final Duration DEFAULT_TIMEOUT_DURATION = Duration.ofMillis(0);
+
+  private String type = "delay";
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/FaultInjectionPolicy.java:25]: declare a constant

File: governance/src/main/java/org/apache/servicecomb/governance/policy/FaultInjectionPolicy.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    this.errorCode = errorCode;
+  }
+
+  public long getDelayTimeToMillis() {
+    return Duration.parse(delayTime).toMillis();
+  }
+
+  @Override
+  public boolean isValid() {
+    if (getDelayTimeToMillis() < 0 && "delay".equals(type)) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/FaultInjectionPolicy.java:71]: use constant

File: governance/src/main/java/org/apache/servicecomb/governance/policy/FaultInjectionPolicy.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      return false;
+    }
+    if ((getErrorCode() < 200 || getErrorCode() > 600) && "abort".equals(type)) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/governance/policy/FaultInjectionPolicy.java:74]: use constant

File: governance/src/main/java/org/apache/servicecomb/injection/DelayFault.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      LOGGER.debug("Fault injection: delay percentage is not configured");
+      return false;
+    }
+    // check fault delay condition.
+    return FaultInjectionUtil.isFaultNeedToInject(param.getReqCount(), delayPercent);
+  }
+
+  @Override
+  public String getName() {
+    return "delay";
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/injection/DelayFault.java:71]: a lot of places use this literal, consider use constant

File: governance/src/main/java/org/apache/servicecomb/injection/FaultResponse.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+  public void setErrorMsg(String errorMsg) {
+    this.errorMsg = errorMsg;
+  }
+
+  public boolean isDelay() {
+    return delay;
+  }
+
+  public void setDelay(boolean delay) {
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/injection/FaultResponse.java:55]: setDelay is never used, can be deleted

File: governance/src/main/java/org/apache/servicecomb/injection/Fault.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+public interface Fault {
+  static <T> CheckedFunction0<T> decorateCheckedSupplier(Fault fault, CheckedFunction0<T> supplier) {
+    return () -> {
+      try {
+        fault.injectFault();
+      } catch (FaultInjectionException exception) {
+        if (!exception.getFaultResponse().isDelay()) {
+          throw exception;
+        }
```
Comment (liubao68) [inline governance/src/main/java/org/apache/servicecomb/injection/Fault.java:30]: This check is useless. DelayFault will not throw FaultInjectionException

**Review by liubao68 — APPROVED** [review]

**Pull request merged**